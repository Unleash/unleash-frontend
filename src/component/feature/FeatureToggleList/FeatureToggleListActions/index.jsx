import FeatureToggleListActions from "./FeatureToggleListActions";

export default FeatureToggleListActions;
