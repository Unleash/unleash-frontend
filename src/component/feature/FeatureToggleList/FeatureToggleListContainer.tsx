import { useFeatures } from 'hooks/api/getters/useFeatures/useFeatures';
import { FeatureSchema } from 'openapi';
import { FeatureToggleListTable } from './FeatureToggleListTable/FeatureToggleListTable';

const featuresPlaceholder: FeatureSchema[] = Array(15).fill({
    name: 'Name of the feature',
    description: 'Short description of the feature',
    type: '-',
    createdAt: new Date(2022, 1, 1),
    project: 'projectID',
});

export const FeatureToggleListContainer = () => {
    const { features = [], loading } = useFeatures();

    return (
        <FeatureToggleListTable
            data={loading ? featuresPlaceholder : features}
            isLoading={loading}
        />
    );
};
