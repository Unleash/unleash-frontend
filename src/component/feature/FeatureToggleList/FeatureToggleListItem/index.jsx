import FeatureToggleListItem from "./FeatureToggleListItem";

export default FeatureToggleListItem;
