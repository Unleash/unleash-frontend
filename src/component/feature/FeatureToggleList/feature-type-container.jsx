import { connect } from 'react-redux';
import Component from './feature-type-component';

const mapStateToProps = state => ({
    types: state.featureTypes.toJS(),
});

const FeatureType = connect(mapStateToProps)(Component);

export default FeatureType;
