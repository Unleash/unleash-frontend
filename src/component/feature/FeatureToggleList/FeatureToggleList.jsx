import { useContext } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { Button, IconButton, List, ListItem, Tooltip } from '@material-ui/core';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { Add } from '@material-ui/icons';

import FeatureToggleListItem from './FeatureToggleListItem';
import SearchField from '../../common/SearchField/SearchField';
import FeatureToggleListActions from './FeatureToggleListActions';
import ConditionallyRender from '../../common/ConditionallyRender/ConditionallyRender';
import PageContent from '../../common/PageContent/PageContent';
import HeaderTitle from '../../common/HeaderTitle';

import loadingFeatures from './loadingFeatures';

import { CREATE_FEATURE } from '../../providers/AccessProvider/permissions';

import AccessContext from '../../../contexts/AccessContext';

import { useStyles } from './styles';
import ListPlaceholder from '../../common/ListPlaceholder/ListPlaceholder';
import { getCreateTogglePath } from '../../../utils/route-path-helpers';
import { NAVIGATE_TO_CREATE_FEATURE } from '../../../testIds';
import { resolveFilteredProjectId } from '../../../hooks/useFeaturesFilter';

const FeatureToggleList = ({
    features,
    revive,
    archive,
    loading,
    flags,
    filter,
    setFilter,
    sort,
    setSort,
}) => {
    const { hasAccess } = useContext(AccessContext);
    const styles = useStyles();
    const smallScreen = useMediaQuery('(max-width:800px)');
    const mobileView = useMediaQuery('(max-width:600px)');

    const setFilterQuery = v => {
        const query = v && typeof v === 'string' ? v.trim() : '';
        setFilter(prev => ({ ...prev, query }));
    };

    const resolvedProjectId = resolveFilteredProjectId(filter);
    const createURL = getCreateTogglePath(resolvedProjectId, flags.E);

    const renderFeatures = () => {
        features.forEach(e => {
            e.reviveName = e.name;
        });

        if (loading) {
            return loadingFeatures.map(feature => (
                <FeatureToggleListItem
                    key={feature.name}
                    feature={feature}
                    revive={revive}
                    hasAccess={hasAccess}
                    className={'skeleton'}
                    flags={flags}
                />
            ));
        }

        return (
            <ConditionallyRender
                condition={features.length > 0}
                show={features.map(feature => (
                    <FeatureToggleListItem
                        key={feature.name}
                        feature={feature}
                        revive={revive}
                        hasAccess={hasAccess}
                        flags={flags}
                    />
                ))}
                elseShow={
                    <ConditionallyRender
                        condition={archive}
                        show={
                            <ListItem className={styles.emptyStateListItem}>
                                No archived features.
                            </ListItem>
                        }
                        elseShow={
                            <ListPlaceholder
                                text="No features available. Get started by adding a
                                new feature toggle."
                                link={createURL}
                                linkText="Add your first toggle"
                            />
                        }
                    />
                }
            />
        );
    };

    const headerTitle = archive ? 'Archived Features' : 'Features';

    return (
        <div className={styles.featureContainer}>
            <div className={styles.searchBarContainer}>
                <SearchField
                    initialValue={filter.query}
                    updateValue={setFilterQuery}
                    className={classnames(styles.searchBar, {
                        skeleton: loading,
                    })}
                />
                <ConditionallyRender
                    condition={!mobileView && !archive}
                    show={<Link to="/archive">Archive</Link>}
                />
            </div>

            <PageContent
                headerContent={
                    <HeaderTitle
                        loading={loading}
                        title={headerTitle}
                        actions={
                            <div className={styles.actionsContainer}>
                                <ConditionallyRender
                                    condition={!smallScreen}
                                    show={
                                        <FeatureToggleListActions
                                            filter={filter}
                                            setFilter={setFilter}
                                            sort={sort}
                                            setSort={setSort}
                                            loading={loading}
                                        />
                                    }
                                />
                                <ConditionallyRender
                                    condition={!archive}
                                    show={
                                        <ConditionallyRender
                                            condition={smallScreen}
                                            show={
                                                <Tooltip title="Create feature toggle">
                                                    <IconButton
                                                        component={Link}
                                                        to={createURL}
                                                        data-test={
                                                            NAVIGATE_TO_CREATE_FEATURE
                                                        }
                                                        disabled={
                                                            !hasAccess(
                                                                CREATE_FEATURE,
                                                                resolvedProjectId
                                                            )
                                                        }
                                                    >
                                                        <Add />
                                                    </IconButton>
                                                </Tooltip>
                                            }
                                            elseShow={
                                                <Button
                                                    to={createURL}
                                                    color="primary"
                                                    variant="contained"
                                                    component={Link}
                                                    data-test={
                                                        NAVIGATE_TO_CREATE_FEATURE
                                                    }
                                                    disabled={
                                                        !hasAccess(
                                                            CREATE_FEATURE,
                                                            resolvedProjectId
                                                        )
                                                    }
                                                    className={classnames({
                                                        skeleton: loading,
                                                    })}
                                                >
                                                    Create feature toggle
                                                </Button>
                                            }
                                        />
                                    }
                                />
                            </div>
                        }
                    />
                }
            >
                <List>{renderFeatures()}</List>
            </PageContent>
        </div>
    );
};

FeatureToggleList.propTypes = {
    features: PropTypes.array.isRequired,
    revive: PropTypes.func,
    loading: PropTypes.bool,
    archive: PropTypes.bool,
    flags: PropTypes.object,
    filter: PropTypes.object.isRequired,
    setFilter: PropTypes.func.isRequired,
    sort: PropTypes.object.isRequired,
    setSort: PropTypes.func.isRequired,
};

export default FeatureToggleList;
