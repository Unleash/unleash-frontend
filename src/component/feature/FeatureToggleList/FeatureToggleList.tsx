import { Dispatch, SetStateAction, useContext, VFC } from 'react';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { List, ListItem } from '@material-ui/core';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import { IFlags } from 'interfaces/uiConfig';
import { SearchField } from 'component/common/SearchField/SearchField';
import { ConditionallyRender } from 'component/common/ConditionallyRender/ConditionallyRender';
import PageContent from 'component/common/PageContent/PageContent';
import { HeaderTitle } from 'component/common/HeaderTitle/HeaderTitle';
import AccessContext from 'contexts/AccessContext';
import ListPlaceholder from 'component/common/ListPlaceholder/ListPlaceholder';
import { IFeaturesFilter } from 'hooks/useFeaturesFilter';
import { FeatureToggleListItem } from './FeatureToggleListItem/FeatureToggleListItem';
import { FeatureToggleListActions } from './FeatureToggleListActions/FeatureToggleListActions';
import { CreateFeatureButton } from '../CreateFeatureButton/CreateFeatureButton';
import { useCreateFeaturePath } from '../CreateFeatureButton/useCreateFeaturePath';
import { IFeaturesSort } from 'hooks/useFeaturesSort';
import { FeatureSchema } from 'openapi';
import { useStyles } from './styles';

interface IFeatureToggleListProps {
    features: FeatureSchema[];
    loading?: boolean;
    flags?: IFlags;
    filter: IFeaturesFilter;
    setFilter: Dispatch<SetStateAction<IFeaturesFilter>>;
    sort: IFeaturesSort;
    setSort: Dispatch<SetStateAction<IFeaturesSort>>;
    onRevive?: (feature: string) => void;
    inProject?: boolean;
    isArchive?: boolean;
}

const loadingFeaturesPlaceholder: FeatureSchema[] = Array(10)
    .fill({
        createdAt: '2021-03-19T09:16:21.329Z',
        description: ' ',
        enabled: true,
        lastSeenAt: '2021-03-24T10:46:38.036Z',
        name: '',
        project: 'default',
        stale: true,
        strategies: [],
        variants: [],
        type: 'release',
        archived: false,
        environments: [],
        impressionData: false,
    })
    .map((feature, index) => ({ ...feature, name: `${index}` })); // ID for React key

export const FeatureToggleList: VFC<IFeatureToggleListProps> = ({
    features,
    onRevive,
    inProject,
    isArchive,
    loading,
    flags,
    filter,
    setFilter,
    sort,
    setSort,
}) => {
    const { hasAccess } = useContext(AccessContext);
    const createFeature = useCreateFeaturePath(filter);
    const styles = useStyles();
    const smallScreen = useMediaQuery('(max-width:800px)');
    const mobileView = useMediaQuery('(max-width:600px)');

    const setFilterQuery = (v: string) => {
        const query = v && typeof v === 'string' ? v.trim() : '';
        setFilter(prev => ({ ...prev, query }));
    };

    const renderFeatures = () => {
        if (loading) {
            return loadingFeaturesPlaceholder.map(feature => (
                <FeatureToggleListItem
                    key={feature.name}
                    feature={feature}
                    onRevive={onRevive}
                    hasAccess={hasAccess}
                    className={'skeleton'}
                    flags={flags}
                />
            ));
        }

        return (
            <ConditionallyRender
                condition={features.length > 0}
                show={features.map(feature => (
                    <FeatureToggleListItem
                        key={feature.name}
                        feature={feature}
                        onRevive={onRevive}
                        hasAccess={hasAccess}
                        flags={flags}
                        inProject={inProject}
                    />
                ))}
                elseShow={
                    <ConditionallyRender
                        condition={Boolean(isArchive)}
                        show={
                            <ListItem className={styles.emptyStateListItem}>
                                No archived features.
                            </ListItem>
                        }
                        elseShow={
                            <ConditionallyRender
                                condition={Boolean(createFeature?.access)}
                                show={() => (
                                    <ListPlaceholder
                                        text="No features available. Get started by adding a new feature toggle."
                                        link={createFeature?.path}
                                        linkText="Add your first toggle"
                                    />
                                )}
                            />
                        }
                    />
                }
            />
        );
    };

    const searchResultsHeader = filter.query
        ? ` (${features.length} matches)`
        : '';

    const headerTitle = isArchive
        ? inProject
            ? `Project archived features ${searchResultsHeader}`
            : `Archived Features ${searchResultsHeader}`
        : `Features ${searchResultsHeader}`;

    return (
        <div>
            <div
                className={classnames(styles.searchBarContainer, {
                    dense: inProject,
                })}
            >
                <SearchField
                    initialValue={filter.query}
                    updateValue={setFilterQuery}
                    showValueChip={!mobileView}
                    className={classnames(styles.searchBar, {
                        skeleton: loading,
                    })}
                />
                <ConditionallyRender
                    condition={!mobileView && !isArchive}
                    show={<Link to="/archive">Archive</Link>}
                />
            </div>

            <PageContent
                headerContent={
                    <HeaderTitle
                        loading={loading}
                        title={headerTitle}
                        actions={
                            <div className={styles.actionsContainer}>
                                <ConditionallyRender
                                    condition={!smallScreen}
                                    show={
                                        <FeatureToggleListActions
                                            filter={filter}
                                            setFilter={setFilter}
                                            sort={sort}
                                            setSort={setSort}
                                            loading={loading}
                                            inProject={inProject}
                                        />
                                    }
                                />
                                <ConditionallyRender
                                    condition={!isArchive}
                                    show={
                                        <CreateFeatureButton
                                            filter={filter}
                                            loading={Boolean(loading)}
                                        />
                                    }
                                />
                            </div>
                        }
                    />
                }
            >
                <List>{renderFeatures()}</List>
            </PageContent>
        </div>
    );
};
