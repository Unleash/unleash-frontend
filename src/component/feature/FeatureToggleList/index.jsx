import { connect } from 'react-redux';
import { toggleFeature, fetchFeatureToggles } from '../../../store/feature-toggle/actions';
import { updateSettingForGroup } from '../../../store/settings/actions';
import FeatureToggleList from './FeatureToggleList';

import { hasPermission } from '../../../permissions';

function checkConstraints(strategy, regex) {
    if (!strategy.constraints) {
        return;
    }
    return strategy.constraints.some(c => c.values.some(v => regex.test(v)));
}

export const mapStateToPropsConfigurable = isFeature => state => {
    const featureMetrics = state.featureMetrics.toJS();
    const settings = state.settings.toJS().feature || {};
    let features = isFeature ? state.features.toJS() : state.archive.get('list').toArray();

    if (settings.currentProjectId && settings.currentProjectId !== '*') {
        features = features.filter(f => f.project === settings.currentProjectId);
    }

    if (settings.filter) {
        try {
            const regex = new RegExp(settings.filter, 'i');
            features = features.filter(
                feature =>
                    feature.strategies.some(s => checkConstraints(s, regex)) ||
                    regex.test(feature.name) ||
                    regex.test(feature.description) ||
                    feature.strategies.some(s => s && s.name && regex.test(s.name)) ||
                    (settings.filter.length > 1 && regex.test(JSON.stringify(feature)))
            );
        } catch (e) {
            // Invalid filter regex
        }
    }

    if (!settings.sort) {
        settings.sort = 'name';
    }

    if (settings.sort === 'enabled') {
        features = features.sort((a, b) =>
            // eslint-disable-next-line
            a.enabled === b.enabled ? 0 : a.enabled ? -1 : 1
        );
    } else if (settings.sort === 'stale') {
        features = features.sort((a, b) =>
            // eslint-disable-next-line
            a.stale === b.stale ? 0 : a.stale ? -1 : 1
        );
    } else if (settings.sort === 'created') {
        features = features.sort((a, b) => (new Date(a.createdAt) > new Date(b.createdAt) ? -1 : 1));
    } else if (settings.sort === 'Last seen') {
        features = features.sort((a, b) => (new Date(a.lastSeenAt) > new Date(b.lastSeenAt) ? -1 : 1));
    } else if (settings.sort === 'name') {
        features = features.sort((a, b) => {
            if (a.name < b.name) {
                return -1;
            }
            if (a.name > b.name) {
                return 1;
            }
            return 0;
        });
    } else if (settings.sort === 'strategies') {
        features = features.sort((a, b) => (a.strategies.length > b.strategies.length ? -1 : 1));
    } else if (settings.sort === 'type') {
        features = features.sort((a, b) => {
            if (a.type < b.type) {
                return -1;
            }
            if (a.type > b.type) {
                return 1;
            }
            return 0;
        });
    } else if (settings.sort === 'metrics') {
        const target = settings.showLastHour ? featureMetrics.lastHour : featureMetrics.lastMinute;

        features = features.sort((a, b) => {
            if (!target[a.name]) {
                return 1;
            }
            if (!target[b.name]) {
                return -1;
            }
            if (target[a.name].yes > target[b.name].yes) {
                return -1;
            }
            return 1;
        });
    }

    return {
        features,
        featureMetrics,
        settings,
        hasPermission: hasPermission.bind(null, state.user.get('profile')),
    };
};
const mapStateToProps = mapStateToPropsConfigurable(true);
const mapDispatchToProps = {
    toggleFeature,
    fetcher: () => fetchFeatureToggles(),
    updateSetting: updateSettingForGroup('feature'),
};

const FeatureToggleListContainer = connect(mapStateToProps, mapDispatchToProps)(FeatureToggleList);

export default FeatureToggleListContainer;
