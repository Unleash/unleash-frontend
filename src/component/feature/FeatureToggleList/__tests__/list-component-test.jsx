import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/core';

import FeatureToggleList from '../FeatureToggleList';
import renderer from 'react-test-renderer';
import theme from '../../../../themes/main-theme';
import { createFakeStore } from '../../../../accessStoreFake';
import {
    ADMIN,
    CREATE_FEATURE,
} from '../../../providers/AccessProvider/permissions';
import AccessProvider from '../../../providers/AccessProvider/AccessProvider';

jest.mock('../FeatureToggleListItem', () => ({
    __esModule: true,
    default: 'ListItem',
}));

jest.mock('../../../common/ProjectSelect');

test('renders correctly with one feature', () => {
    const features = [
        {
            name: 'Another',
        },
    ];

    const tree = renderer.create(
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <AccessProvider
                    store={createFakeStore([{ permission: CREATE_FEATURE }])}
                >
                    <FeatureToggleList
                        updateSetting={jest.fn()}
                        filter={{}}
                        setFilter={jest.fn()}
                        sort={{}}
                        setSort={jest.fn()}
                        features={features}
                        fetcher={jest.fn()}
                        flags={{}}
                    />
                </AccessProvider>
            </ThemeProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});

test('renders correctly with one feature without permissions', () => {
    const features = [
        {
            name: 'Another',
        },
    ];
    const tree = renderer.create(
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <AccessProvider
                    store={createFakeStore([{ permission: CREATE_FEATURE }])}
                >
                    <FeatureToggleList
                        filter={{}}
                        setFilter={jest.fn()}
                        sort={{}}
                        setSort={jest.fn()}
                        features={features}
                        fetcher={jest.fn()}
                        flags={{}}
                    />
                </AccessProvider>
            </ThemeProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
