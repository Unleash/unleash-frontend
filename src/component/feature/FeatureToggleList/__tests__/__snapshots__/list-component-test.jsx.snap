// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with one feature 1`] = `
<div
  className="featureContainer"
>
  <div
    className="searchBarContainer"
  >
    <div>
      <div
        className="MuiFormControl-root MuiTextField-root"
        onKeyPress={[Function]}
        style={
          Object {
            "maxWidth": "80%",
            "width": "500px",
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
          data-shrink={false}
        >
          Search
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Search
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge"
      data-test="add-feature-btn"
      href="/features/create"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiButton-label"
      >
        Add new feature
      </span>
    </a>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    shadow={0}
  >
    <div>
      <div
        className="listComponentHeader"
      >
        <button
          aria-controls="metric"
          aria-haspopup="true"
          className="MuiButtonBase-root MuiButton-root MuiButton-text dropdownButton"
          disabled={false}
          id="metric"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Metric interval"
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Last minute
            <span
              className="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                arrow_drop_down
              </span>
            </span>
          </span>
        </button>
        <button
          aria-controls="sorting"
          aria-haspopup="true"
          className="MuiButtonBase-root MuiButton-root MuiButton-text dropdownButton"
          disabled={false}
          id="sorting"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Sort by"
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            By name
            <span
              className="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                arrow_drop_down
              </span>
            </span>
          </span>
        </button>
        <Project
          settings={
            Object {
              "sort": "name",
            }
          }
          updateSetting={[MockFunction]}
        />
      </div>
    </div>
    <hr />
    <ul
      className="MuiList-root MuiList-padding"
    >
      <ListItem
        feature={
          Object {
            "name": "Another",
            "reviveName": "Another",
          }
        }
        hasPermission={[Function]}
        setFilter={[Function]}
        settings={
          Object {
            "sort": "name",
          }
        }
        toggleFeature={[MockFunction]}
      />
    </ul>
  </div>
</div>
`;

exports[`renders correctly with one feature without permissions 1`] = `
<div
  className="featureContainer"
>
  <div
    className="searchBarContainer"
  >
    <div>
      <div
        className="MuiFormControl-root MuiTextField-root"
        onKeyPress={[Function]}
        style={
          Object {
            "maxWidth": "80%",
            "width": "500px",
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
          data-shrink={false}
        >
          Search
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Search
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    shadow={0}
  >
    <div>
      <div
        className="listComponentHeader"
      >
        <button
          aria-controls="metric"
          aria-haspopup="true"
          className="MuiButtonBase-root MuiButton-root MuiButton-text dropdownButton"
          disabled={false}
          id="metric"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Metric interval"
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Last minute
            <span
              className="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                arrow_drop_down
              </span>
            </span>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-controls="sorting"
          aria-haspopup="true"
          className="MuiButtonBase-root MuiButton-root MuiButton-text dropdownButton"
          disabled={false}
          id="sorting"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Sort by"
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            By name
            <span
              className="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <span
                aria-hidden={true}
                className="material-icons MuiIcon-root"
              >
                arrow_drop_down
              </span>
            </span>
          </span>
          <span
            className="MuiTouchRipple-root"
          />
        </button>
        <Project
          settings={
            Object {
              "sort": "name",
            }
          }
          updateSetting={[MockFunction]}
        />
      </div>
    </div>
    <hr />
    <ul
      className="MuiList-root MuiList-padding"
    >
      <ListItem
        feature={
          Object {
            "name": "Another",
            "reviveName": "Another",
          }
        }
        hasPermission={[Function]}
        setFilter={[Function]}
        settings={
          Object {
            "sort": "name",
          }
        }
        toggleFeature={[MockFunction]}
      />
    </ul>
  </div>
</div>
`;
